### PR TITLE
Often ignore hard clipping when fetching unclipped ends

### PR DESCRIPTION
**Pull request by clintval**
Motivated by:

- https://github.com/fulcrumgenomics/fgbio/issues/1030

Closes https://github.com/fulcrumgenomics/fgbio/issues/878
Closes https://github.com/fulcrumgenomics/fgbio/pull/937
Closes https://github.com/fulcrumgenomics/fgbio/issues/1030

_**Complete PR justification comment coming soon, opening as draft for now...**_

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/bam/api/SamRecord.scala [hunk 1]
```diff
@@ -186,13 +189,30 @@ trait SamRecord {
...
+      mateStart - cigar.leadingSoftClippedBases
+    }
+  }
+  @inline final def mateUnSoftClippedEnd: Option[Int] = {
+    require(paired && mateMapped, "Cannot get mate un-soft-clipped end position on read without a mapped mate.")
+    get[String]("MC").map { cig =>
+      val cigar = Cigar(cig)
+      mateStart + cigar.lengthOnTarget - 1 + cigar.trailingSoftClippedBases
     }
   }
```
Comment (clintval) [inline src/main/scala/com/fulcrumgenomics/bam/api/SamRecord.scala:217]: I'm starting to think we should cache the mate cigar like we do with the record's cigar. It would require similar trickery to detect when a new mate cigar is updated upon the record since we would have to intercept any call to `setAttribute`, check if the attribute is being set with the `MC` tag, and then reset the in-memory mate cigar to the new value. If I implement this suggestion, it will be in a separate PR.

File: src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala [hunk 2]
```diff
@@ -327,23 +327,23 @@ class SamRecordClipper(val mode: ClippingMode, val autoClipAttributes: Boolean)
   /** Returns the number of bases extending past the mate end for FR pairs including any soft-clipped bases, zero otherwise.
     *
     * @param rec the record to examine
-    * @param mateUnclippedStart the smallest mapped genomic coordinate considering clipping
-    * @param mateUnclippedEnd the largest mapped genomic coordinate considering clipping
+    * @param mateUnSoftClippedStart the smallest mapped genomic coordinate considering *ONLY* soft-clipping
+    * @param mateUnSoftClippedEnd the largest mapped genomic coordinate considering *ONLY* soft-clipping
```
Comment (clintval) [inline src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala:331]: I don't love the new names for these but really couldn't think of anything better.

File: src/test/scala/com/fulcrumgenomics/bam/ClipBamTest.scala
Comment (clintval) [inline src/test/scala/com/fulcrumgenomics/bam/ClipBamTest.scala:1]: No old tests need to be updated which makes me feel confident this edge case hasn't been considered or explored much yet.

File: src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala [hunk 3]
```diff
@@ -287,12 +287,12 @@ trait UmiConsensusCaller[ConsensusRead <: SimpleRead] {
       var index = if (!rec.isFrPair) trimToLength - 1 else {
         // Get the number of mapped bases to clip that maps beyond the mate's end, including any soft-clipped bases. Use
         // that to compute where in the read to keep.
-        val mateUnclippedStart = rec.mateUnclippedStart.getOrElse(throw new IllegalArgumentException(f"Mate cigar (MC SAM tag) needed for read: ${rec.name}"))
-        val mateUnclippedEnd   = rec.mateUnclippedEnd.getOrElse(throw new IllegalArgumentException(f"Mate cigar (MC SAM tag) needed for read: ${rec.name}"))
+        val mateUnSoftClippedStart = rec.mateUnSoftClippedStart.getOrElse(throw new IllegalArgumentException(f"Mate cigar (MC SAM tag) needed for read: ${rec.name}"))
+        val mateUnSoftClippedEnd   = rec.mateUnSoftClippedEnd.getOrElse(throw new IllegalArgumentException(f"Mate cigar (MC SAM tag) needed for read: ${rec.name}"))
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala]: [suggestion(if-minor)](https://pullpo.io/cc?l=suggestion&d=if-minor) 
consider a note or TODO that the MC is parsed twice

Comment (clintval) [reply src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala]: I will open an issue!

This will be fixed when the mate cigar can be cached. I'm going to open an issue to track that change so I don't need to add a comment to the `UmiConsensusCaller` at this time. The double-parse also existed before and after this PR too.

File: src/main/scala/com/fulcrumgenomics/bam/api/SamRecord.scala (hunk 1 above)
Comment (clintval) [reply src/main/scala/com/fulcrumgenomics/bam/api/SamRecord.scala]: https://github.com/fulcrumgenomics/fgbio/issues/1115
